### ISSUE TITLE
TAS classifier `predict` requires hardcoded columns in input DataFrame [BUG]

### ISSUE THREAD

**Issue by mtb-za** [morganjwilliams/pyrolite#48]
**Describe the bug**
The `utils.classification.TAS` has a `predict` method, to assign a label from the TAS scheme. This only works if the DataFrame has the specific columns `'SiO2'` and `'Na2O + K2O'` present, giving an error (`"None of [Index(['sio2', 'na2o + k2o'], dtype='object')] are in the [columns]"`) if they are not, or are spelled/capitalised differently. (My dataset has all the oxides in lowercase, for example.)

**To Reproduce**
1. Use the guide from https://pyrolite.readthedocs.io/en/develop/examples/util/TAS.html#sphx-glr-examples-util-tas-py
2. Change (or have) the relevant columns as lower-case (eg: `df.rename({'SiO2': 'sio2', 'Na2O + K2O': 'na2o + k2o'})`
3. `df["TAS"] = cm.predict(df)` fails.
4. 

**Expected behavior**
This should do the prediction regardless of the case of the column names. Ideally, it should also work if the names are different for some reason. If `'Na2O + K2O'` is named `'total_alkali'` for example, then it should still be able to use those values. This can be done by using the current as a default, but doing a caseless comparison, as well as allowing for the two columns to be passed to the `predict` method somewhere.

**Environment:**
 - OS: Windows 10
 - Python Version 3.8.5
 - Pyrolite Version 0.2.8

**Additional Context**
Happy to look at making a PR for this, if you think it worthwhile.

**Comment by morganjwilliams** [comment]
Hi @mtb-za, thanks for the bug report!

**TL;DR:** In this case I think the simplest scenario is to just pass a numpy array to `cm.predict`, which in your case would be `cm.predict(df['sio2', 'na2o + k2o'].values)`. This is going to be fairly similar to the alternative you note where the TAS classifier itself handles these names (in content and length), but without any needed change.

The behavior of the TAS classifier is largely inherited from the [`PolygonClassifer`](https://pyrolite.readthedocs.io/en/develop/api/util/classification.html#pyrolite.util.classification.PolygonClassifier), which is defined with a set of `axes` (variables, not a plot) which it then uses for i) the prediction (where a dataframe is passed) and ii) for axes labels. As of a small bugfix on **develop** in e29035740efd0afa8799d3c326f73f7ae43bfd1e (there was indeed a true bug in TAS, but different to the issue you describe!), this can be directly passed to `TAS()`. If for some reason you don't want to pass a numpy array as I note above, you could install the development version and instantiate your TAS classifier with your locally used variable names (`cm = TAS(axes=['sio2', 'na2o + k2o'])`) and go from there. This would then allow you to just pass the whole dataframe subsequently and the classifier will pull out what it needs (`cm.predict(df)` should then work as expected for your dataframe/dataframes). This is how I'd generally intended `PolygonClassifer` to be used; perhaps there's some scope for extending it a little (e.g. for TAS if it doesn't find TotalAlkali or Na2O + K2O but it does find Na2O and K2O it's easy enough to construct).

For the moment a decent amount of the pandas-based variable handling throughout pyrolite is dependent on properly cased element and mineral formulae; might be able to work on some different "modes" for this in the future but that's a bit of a larger project (along with dealing with mass balance and handling ionic species). If this constraint is removed there are a number of edge cases where things can't be disambiguated (e.g. CO, Co, co), and we'll need a system for warning users that's the case (lest it pass silently...). There is a function which is intended to properly case most element, oxide and isotope ratio column names ([`pyrolite.geochem.parse.tochem`](https://pyrolite.readthedocs.io/en/develop/api/geochem/parse.html#pyrolite.geochem.parse.tochem)), and this would convert both lower-cased and upper-cased formulae to appropriate case - but doesn't currently have any warning to deal with the edge cases I mentioned.

Let me know if this is a viable solution for you, or if you still think there's a change needed/I've missed something.

**Comment by mtb-za** [comment]
Thanks for the detailed response. I think that both approaches that you outline are a sensible workaround.

The last time that I was doing geochem/pretrology was a while back, so I am not surprised that there are edge cases where case does matter; I was just a little surprised at the rigidity here. The `parse.tochem` is also a nice approach, which I missed on looking through the docs at first.

**Comment by morganjwilliams** [comment]
No worries! For most of the things in pyrolite, there are at least two ways to go about things. I've tried to make things accept numpy arrays where possible to deal with scenarios where you don't want to deal with the builtin functionality based on dataframes (and the rigidity which sometimes comes with it); in most cases I think the flexibility is useful and makes it a bit more reusable/extensible, in a few it's perhaps more confusing than it's worth.

If you're happy with the workaround I'll close this one for now, but will likely reference it in future when plans to deal with a few of the issues mentioned above come to fruition.

**Comment by mtb-za** [comment]
Sounds good to me. Thanks for the help. :)
